File: ios/GradeThread/Capture/PhotoIntakeStore.swift
import Foundation
import Observation

/// Owns the per-session capture state for the photo intake flow.
/// Survives the camera lifecycle (start/stop) but is destroyed when the
/// user leaves the intake view.
///
/// US-2470: slots are ``CaptureSlot`` — a (photo_type, photo_role) pair — and
/// which ones exist comes from the resolved ``PhotoProfile``, not from a fixed
/// list of enum cases. A suit profile can therefore hold three separate tag
/// slots, and a bottom is never offered a sleeve measurement.
@MainActor
@Observable
public final class PhotoIntakeStore {
    /// Captured photos keyed by slot. A missing key = empty slot.
    public private(set) var photos: [CaptureSlot: PhotoCapture] = [:]

    /// Slot the next capture will land in. Capture-flow swipes / taps
    /// update this; the auto-advance after each capture also writes here.
    public var activeSlot: CaptureSlot = CaptureSlot(.front)

    /// The resolved photo profile for this item's category. Drives which slots
    /// the strip shows, what they are called, and their `sort_order`.
    ///
    /// Starts on the bundled clothing fallback so the strip is never empty
    /// while ``PhotoProfileStore`` fetches the server table.
    public private(set) var profile: PhotoProfile = .clothingFallback

    /// Slots always shown in the strip, in profile order.
    public private(set) var baseSlots: [CaptureSlot] = PhotoProfile.clothingFallback.defaultCaptureSlots

    /// Optional slots the user has revealed, in reveal order. Defects reveal
    /// one at a time via ``revealNextDefectSlot()``; every other optional slot
    /// reveals directly via ``reveal(_:)`` from the "Add more" menu.
    public private(set) var extraSlots: [CaptureSlot] = []

    public init() {}

    // MARK: - Profile

    /// Swaps in a resolved profile.
    ///
    /// Anything already CAPTURED under a slot the new profile does not show by
    /// default is moved into ``extraSlots`` rather than dropped. The server
    /// table lands mid-session (the fetch races the camera opening), and a
    /// seller who had already shot a tag would otherwise watch it vanish from
    /// the strip while its bytes sat in `photos` waiting to upload.
    public func apply(profile newProfile: PhotoProfile) {
        guard newProfile != profile else { return }
        // Snapshot the OLD strip order before anything moves. `photos.keys` is
        // unordered, so carrying slots over from the dictionary alone would
        // shuffle the strip differently on each launch.
        let oldOrder = visibleSlots
        profile = newProfile
        let newBase = newProfile.defaultCaptureSlots
        baseSlots = newBase

        var carried: [CaptureSlot] = []
        for slot in oldOrder where !newBase.contains(slot) && !carried.contains(slot) {
            if extraSlots.contains(slot) || photos[slot] != nil { carried.append(slot) }
        }
        for slot in photos.keys where !newBase.contains(slot) && !carried.contains(slot) {
            carried.append(slot)
        }
        extraSlots = carried

        if !visibleSlots.contains(activeSlot) {
            activeSlot = nextEmptySlot ?? visibleSlots.first ?? CaptureSlot(.front)
        }
    }

    // MARK: - Derived state

    /// Slots currently displayed in the strip: the profile's default slots plus
    /// whatever the seller has revealed.
    public var visibleSlots: [CaptureSlot] {
        baseSlots + extraSlots
    }

    /// How many defect slots are currently revealed (0…3).
    public var defectSlotsVisible: Int {
        extraSlots.filter { $0.isDefect }.count
    }

    /// First slot without a capture, scanning the visible strip in order.
    /// Used to auto-advance after a capture lands.
    public var nextEmptySlot: CaptureSlot? {
        visibleSlots.first { photos[$0] == nil }
    }

    /// The blocking slots for this profile — front + back in every profile
    /// shipped so far, but read from the profile rather than assumed.
    public var requiredSlots: [CaptureSlot] {
        let blocking = profile.captureSlots.filter { $0.isBlocking }
        return blocking.isEmpty ? CaptureSlot.blocking : blocking
    }

    public var allRequiredFilled: Bool {
        requiredSlots.allSatisfy { photos[$0] != nil }
    }

    /// True iff the user has any captured photos. Drives the exit-
    /// confirmation prompt — if they haven't snapped anything, the X
    /// dismisses without warning.
    public var hasUnsavedShots: Bool {
        !photos.isEmpty
    }

    public var canAddDefectSlot: Bool {
        nextHiddenDefectSlot != nil
    }

    /// The first defect slot not yet revealed, if any. Empty when the profile
    /// does not offer defect close-ups at all.
    public var nextHiddenDefectSlot: CaptureSlot? {
        profile.defectCaptureSlots.first { !extraSlots.contains($0) }
    }

    /// Optional slots offerable from the "Add more" menu: the next hidden
    /// defect (defects reveal one at a time), then every profile slot not yet
    /// shown, in profile order.
    ///
    /// US-2470: this is the whole point of the story. It used to be
    /// `PhotoSlotType.extras + .measurements` — a fixed list that offered a
    /// pair of trousers an inseam AND a sleeve, and that could only name an
    /// extra tag shot "Tag 2".
    public var hiddenExtraSlots: [CaptureSlot] {
        var out: [CaptureSlot] = []
        if let nextDefect = nextHiddenDefectSlot { out.append(nextDefect) }
        let shown = Set(visibleSlots)
        out += profile.optionalCaptureSlots.filter { !shown.contains($0) }
        return out
    }

    /// The "Add more" entries grouped for the menu: measurements are a section
    /// of their own because a garment can have five of them and they would
    /// otherwise bury everything else.
    public var hiddenMeasurementSlots: [CaptureSlot] {
        hiddenExtraSlots.filter { $0.serverPhotoType == "measurement" }
    }

    /// "Add more" entries that are not measurements and not the defect entry.
    public var hiddenGeneralSlots: [CaptureSlot] {
        hiddenExtraSlots.filter { $0.serverPhotoType != "measurement" && !$0.isDefect }
    }

    // MARK: - Auto-assign (US-2818)

    /// Slots an UNLABELLED batch of photos should land in, in order, so the
    /// seller never has to tag a photo before the item exists.
    ///
    /// This is the web `bulkUpload` rule (photo-uploader.tsx): the first photos
    /// fill whatever REQUIRED slots are still empty — in profile order, so the
    /// required set completes and the item still earns "photographed" — and the
    /// rest land on ordinary listing slots the seller corrects afterwards.
    ///
    /// Two kinds of slot are deliberately never auto-filled, because a wrong tag
    /// on either is worse than no photo at all: DEFECT slots, which tell a buyer
    /// the garment is damaged, and MEASUREMENT slots, whose MeasureCard frame is
    /// a calibration shot the pipeline reads rather than a photo anyone lists.
    /// Both stay available from the tray's per-photo menu.
    public func autoAssignTargets(count: Int) -> [CaptureSlot] {
        guard count > 0 else { return [] }
        func isSafeFiller(_ slot: CaptureSlot) -> Bool {
            !slot.isDefect && slot.serverPhotoType != "measurement"
        }
        var targets: [CaptureSlot] = []
        var taken = Set<CaptureSlot>()
        func offer(_ slot: CaptureSlot) {
            guard targets.count < count, !taken.contains(slot) else { return }
            taken.insert(slot)
            targets.append(slot)
        }
        for slot in requiredSlots where photos[slot] == nil { offer(slot) }
        for slot in visibleSlots where photos[slot] == nil && isSafeFiller(slot) {
            offer(slot)
        }
        for slot in hiddenGeneralSlots { offer(slot) }
        return targets
    }

    // MARK: - Mutations

    /// Stores a photo in the currently-active slot and advances to the
    /// next empty slot (or stays put if everything's filled).
    public func recordCapture(_ photo: PhotoCapture) {
        recordCapture(photo, into: activeSlot)
    }

    /// US-1648: record a capture into an EXPLICIT slot pinned by the caller
    /// BEFORE the async capture, so a slot-strip tap mid-capture/compress can't
    /// redirect the photo into a different slot — e.g. a sensitive tag close-up
    /// landing in the public 'front' slot (and thus the public item-photos
    /// bucket). Advances the focus only when it's still on the captured slot.
    public func recordCapture(_ photo: PhotoCapture, into slot: CaptureSlot) {
        reveal(slot)
        photos[slot] = photo
        if activeSlot == slot, let next = nextEmptySlot {
            activeSlot = next
        }
    }

    /// Stores a photo in a specific slot. Used by the library-import flow
    /// (US-174) which lets the user assign each picked image after picking,
    /// and by draft restore. Optional slots auto-reveal so the assigned photo
    /// is always visible in the strip.
    public func setPhoto(_ photo: PhotoCapture, for slot: CaptureSlot) {
        reveal(slot)
        photos[slot] = photo
    }

    public func clearPhoto(at slot: CaptureSlot) {
        photos.removeValue(forKey: slot)
        // If the cleared slot becomes the natural next-up, focus it so the
        // user can retake without a manual tap.
        if let next = nextEmptySlot, photos[activeSlot] != nil {
            activeSlot = next
        }
    }

    public func setActiveSlot(_ slot: CaptureSlot) {
        // Don't allow focusing an optional slot that hasn't been revealed
        // yet — the slot strip wouldn't be rendering it.
        guard visibleSlots.contains(slot) else { return }
        activeSlot = slot
    }

    /// Reveals an optional slot so it appears in the strip. No-op for the
    /// profile's default slots (always visible) and for already-revealed ones.
    public func reveal(_ slot: CaptureSlot) {
        guard !baseSlots.contains(slot), !extraSlots.contains(slot) else { return }
        extraSlots.append(slot)
    }

    /// Reveals one more defect slot, up to the max of 3.
    public func revealNextDefectSlot() {
        guard let next = nextHiddenDefectSlot else { return }
        reveal(next)
    }

    /// Resets state for a fresh intake session. Keeps the resolved profile —
    /// the category has not changed just because the seller started over.
    public func reset() {
        photos.removeAll()
        extraSlots.removeAll()
        activeSlot = baseSlots.first ?? CaptureSlot(.front)
    }

    // MARK: - Ordering

    /// The captured photos in upload order: profile order first, then anything
    /// the profile does not list (defects, and slots carried over from an
    /// earlier profile) in the order the seller revealed them.
    ///
    /// This IS the `sort_order` contract (US-2470 AC3). It replaces
    /// `PhotoSlotType.allCases.firstIndex(of:)`, which could only express one
    /// global ordering and had no way to put a watch's dial before its
    /// caseback.
    public var orderedCaptures: [(slot: CaptureSlot, capture: PhotoCapture)] {
        let strip = visibleSlots
        func rank(_ slot: CaptureSlot) -> (Int, Int) {
            if let index = profile.sortIndex(of: slot) { return (0, index) }
            return (1, strip.firstIndex(of: slot) ?? Int.max)
        }
        return photos
            .map { (slot: $0.key, capture: $0.value) }
            .sorted {
                let a = rank($0.slot), b = rank($1.slot)
                return a.0 == b.0 ? a.1 < b.1 : a.0 < b.0
            }
    }
}

File: android/app/src/main/java/com/gradethread/app/capture/PhotoIntakeStore.kt
package com.gradethread.app.capture

import com.gradethread.app.sync.db.CaptureDraftEntity
import com.gradethread.app.sync.db.GradeThreadDb
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

/**
 * US-1324: the capture-session state machine (iOS PhotoIntakeStore). Pure
 * over an immutable [State] snapshot so every rule unit-tests; the CameraX
 * screen observes [state].
 *
 * US-2498: a slot is a [CaptureSlot] - a (photo_type, photo_role) pair - and
 * WHICH slots exist comes from the resolved [PhotoProfile], not from a fixed
 * list of enum cases. A suit profile can therefore hold three separate tag
 * shots, and a pair of trousers is never offered a sleeve measurement.
 *
 * Rules carried from iOS:
 *  - the strip shows the profile's default slots + revealed extras; the
 *    profile's blocking slots (front+back) gate continue, the rest are
 *    skippable;
 *  - a capture lands in the ACTIVE slot then AUTO-ADVANCES to the next empty
 *    visible slot (stays put when everything's filled);
 *  - defects reveal ONE AT A TIME (the Add menu offers a single Defect entry
 *    while any remain hidden); every other optional slot reveals on demand;
 *  - the FULL state (photos map + active + revealed slots) persists to Room
 *    so process death/backgrounding recovers the draft.
 */
class PhotoIntakeStore(initial: State = State(), initialProfile: PhotoProfile = PhotoProfile.clothingFallback) {

    @Serializable
    data class State(
        /** [CaptureSlot.storageKey] → captured photo file path. */
        val photos: Map<String, String> = emptyMap(),
        val activeSlot: String = PhotoSlotType.FRONT.wire,
        val extraSlots: List<String> = emptyList(),
    ) {
        fun photoFor(slot: CaptureSlot): String? = photos[slot.storageKey]

        /**
         * The unroled photo for a bare type. A role-less slot's storage key IS
         * its wire value, so this is the same lookup - kept for the surfaces
         * that legitimately deal in types alone (the share-target inbox, which
         * assigns from a picker with no profile in hand).
         */
        fun photoFor(slot: PhotoSlotType): String? = photos[slot.wire]

        val activeCaptureSlot: CaptureSlot
            get() = CaptureSlot.fromStorageKey(activeSlot) ?: CaptureSlot(PhotoSlotType.FRONT)
        val active: PhotoSlotType
            get() = activeCaptureSlot.type
        val revealedSlots: List<CaptureSlot>
            get() = extraSlots.mapNotNull(CaptureSlot::fromStorageKey)
        val revealed: List<PhotoSlotType>
            get() = revealedSlots.map { it.type }
    }

    private val stateFlow = MutableStateFlow(initial)
    val state: StateFlow<State> = stateFlow

    private val profileFlow = MutableStateFlow(initialProfile)

    /**
     * The resolved profile for this item's category. Drives which slots the
     * strip offers; starts on the bundled fallback so the first frame renders
     * while [PhotoProfileStore] fetches the server table.
     */
    val profile: StateFlow<PhotoProfile> = profileFlow

    // ── Profile ──────────────────────────────────────────────────────────────

    /**
     * Swap in a resolved profile.
     *
     * Anything already CAPTURED under a slot the new profile does not show by
     * default is moved into the revealed extras rather than dropped: the photo
     * exists on disk and the seller took it on purpose, so the strip has to keep
     * a place for it even when the new profile has never heard of that slot.
     */
    fun apply(newProfile: PhotoProfile) {
        if (newProfile == profileFlow.value) return
        val current = stateFlow.value
        val base = newProfile.defaultCaptureSlots.toSet()
        val carried =
            (current.revealedSlots + current.photos.keys.mapNotNull(CaptureSlot::fromStorageKey))
                .distinct()
                .filter { it !in base }
                .map { it.storageKey }

        profileFlow.value = newProfile
        var next = current.copy(extraSlots = carried)
        if (visibleSlotsOf(next).none { it.storageKey == next.activeSlot }) {
            val fallback = visibleSlotsOf(next).firstOrNull { next.photoFor(it) == null }
                ?: visibleSlotsOf(next).firstOrNull()
                ?: CaptureSlot(PhotoSlotType.FRONT)
            next = next.copy(activeSlot = fallback.storageKey)
        }
        stateFlow.value = next
    }

    // ── Derived (mirrors iOS computed properties) ────────────────────────────

    private fun visibleSlotsOf(snapshot: State): List<CaptureSlot> =
        profileFlow.value.defaultCaptureSlots + snapshot.revealedSlots

    val visibleSlots: List<CaptureSlot>
        get() = visibleSlotsOf(stateFlow.value)

    val nextEmptySlot: CaptureSlot?
        get() = visibleSlots.firstOrNull { stateFlow.value.photoFor(it) == null }

    /**
     * The blocking slots for this profile - front + back in every profile
     * shipped so far, but READ from the profile rather than assumed.
     */
    val requiredSlots: List<CaptureSlot>
        get() = profileFlow.value.captureSlots.filter { it.isBlocking }
            .ifEmpty { CaptureSlot.blocking }

    val allRequiredFilled: Boolean
        get() = requiredSlots.all { stateFlow.value.photoFor(it) != null }

    /** Drives the exit-confirmation prompt. */
    val hasUnsavedShots: Boolean
        get() = stateFlow.value.photos.isNotEmpty()

    val nextHiddenDefectSlot: CaptureSlot?
        get() {
            val revealed = stateFlow.value.revealedSlots
            return profileFlow.value.defectCaptureSlots.firstOrNull { it !in revealed }
        }

    val canAddDefectSlot: Boolean
        get() = nextHiddenDefectSlot != null

    /**
     * Add-menu entries: the next hidden defect first (one at a time), then every
     * profile slot not already in the strip, in profile order.
     *
     * US-2498: this used to be `PhotoSlotType.extras + .measurements` - a fixed
     * list that offered a handbag the same three extras it offered a t-shirt.
     */
    val hiddenExtraSlots: List<CaptureSlot>
        get() {
            val out = mutableListOf<CaptureSlot>()
            nextHiddenDefectSlot?.let(out::add)
            val shown = visibleSlots.toSet()
            out += profileFlow.value.optionalCaptureSlots.filter { it !in shown }
            return out
        }

    // ── Mutations ────────────────────────────────────────────────────────────

    /**
     * Where the next [count] photos will land, without recording them.
     *
     * US-2639: the library-import path has to choose a RESOLUTION CAP before it
     * processes, and the cap is per-slot — but the slot is only decided by
     * [recordCapture]'s auto-advance, which runs after. So the destinations have
     * to be predicted.
     *
     * This mirrors [recordCapture] exactly: the first photo goes to the CURRENT
     * active slot (not the next empty one — that difference is the whole trap),
     * and each subsequent photo to the next slot still without a photo.
     * `PhotoIntakeStoreTest` records N photos through the real function and
     * asserts they land in exactly these slots, so the mirroring is pinned
     * rather than assumed — a prediction that drifts would silently compress a
     * serial shot at the default cap.
     */
    fun plannedDestinations(count: Int): List<CaptureSlot> {
        val snapshot = stateFlow.value
        val filled = snapshot.photos.keys.toMutableSet()
        val out = mutableListOf<CaptureSlot>()
        var next: CaptureSlot? = CaptureSlot.fromStorageKey(snapshot.activeSlot)
        repeat(count) {
            val slot = next ?: visibleSlotsOf(snapshot).firstOrNull { it.storageKey !in filled }
            if (slot == null) return out
            out += slot
            filled += slot.storageKey
            next = visibleSlotsOf(snapshot).firstOrNull { it.storageKey !in filled }
        }
        return out
    }

    /**
     * Store a capture in the active slot, then auto-advance to next empty.
     *
     * US-2658: [intoSlot] lets a caller PIN the destination before an async
     * hop. A camera shot is filed after a shutter round trip, and until this
     * existed the key was read from `activeSlot` at CALLBACK time — so tapping
     * a different chip while the shutter was in flight filed the photo under
     * the slot the seller had just moved to. The library-import path passes
     * nothing on purpose: a multi-pick fills successive slots by leaning on the
     * auto-advance below, which is the behaviour it wants.
     */
    fun recordCapture(path: String, intoSlot: String? = null) {
        val current = stateFlow.value
        val key = intoSlot ?: current.activeSlot
        val withPhoto = current.copy(photos = current.photos + (key to path))
        stateFlow.value = withPhoto
        nextEmptySlot?.let { stateFlow.value = withPhoto.copy(activeSlot = it.storageKey) }
    }

    fun setPhoto(slot: CaptureSlot, path: String) {
        val current = stateFlow.value
        stateFlow.value = current.copy(photos = current.photos + (slot.storageKey to path))
    }

    fun setPhoto(slot: PhotoSlotType, path: String) = setPhoto(CaptureSlot(slot), path)

    fun clearPhoto(slot: CaptureSlot) {
        val current = stateFlow.value
        stateFlow.value = current.copy(photos = current.photos - slot.storageKey)
    }

    fun setActiveSlot(slot: CaptureSlot) {
        stateFlow.value = stateFlow.value.copy(activeSlot = slot.storageKey)
    }

    fun setActiveSlot(slot: PhotoSlotType) = setActiveSlot(CaptureSlot(slot))

    /** Reveal an optional slot (defect or profile extra) and activate it. */
    fun reveal(slot: CaptureSlot) {
        val current = stateFlow.value
        if (slot.storageKey in current.extraSlots) {
            stateFlow.value = current.copy(activeSlot = slot.storageKey)
            return
        }
        stateFlow.value = current.copy(
            extraSlots = current.extraSlots + slot.storageKey,
            activeSlot = slot.storageKey,
        )
    }

    fun reveal(slot: PhotoSlotType) = reveal(CaptureSlot(slot))

    // ── Draft persistence (AC3) ──────────────────────────────────────────────

    companion object {
        private val json = Json { ignoreUnknownKeys = true }
        const val DRAFT_ID = "active" // one in-flight capture session

        suspend fun restore(db: GradeThreadDb): PhotoIntakeStore {
            val draft = db.captureDrafts().byId(DRAFT_ID) ?: return PhotoIntakeStore()
            val state = runCatching { json.decodeFromString(State.serializer(), draft.stateJson) }
                .getOrDefault(State())
            return PhotoIntakeStore(state)
        }
    }

    suspend fun persist(db: GradeThreadDb, clock: () -> Long = System::currentTimeMillis) {
        db.captureDrafts().upsert(
            CaptureDraftEntity(
                id = DRAFT_ID,
                stateJson = json.encodeToString(State.serializer(), stateFlow.value),
                updatedAt = clock(),
            ),
        )
    }

    /** The intake completed (or was explicitly discarded) — drop the draft. */
    suspend fun discardDraft(db: GradeThreadDb) {
        db.captureDrafts().delete(DRAFT_ID)
    }
}
